Modify middleware listener to use stratigility pipe

asgrim opened the pull request:
Taking issue #196 into play, this PR proposes to update the `MiddlewareListener` to use `Stratigility` to allow piped middleware. This means the `middleware` param can become an `array` and everything is pushed into the `MiddlewarePipe`... as let's be fair, only allowing just a single middleware is only marginally useful.

For this to work, `zendframework/zend-stratigility` needs to become a requirement for those using `MiddlewareListener` (I added it as `require-dev` for now), so could be seen as a BC break potentially; so happy to negotiate on how we'd like this added to `composer.json`. Otherwise, I believe (pending CI results) all the other tests still pass though.

asgrim committed:
Require stratigility in dev, and suggestion

asgrim committed:
Added tests for handling an array of middleware to be piped

asgrim committed:
Added implementation of Stratigility middleware pipe

asgrim commented:
Apparently I didn't have an up to date `develop` branch >.<

asgrim committed:
Updated RotueMatch#getParams calls to be expected in tests

asgrim commented:
Rebased correctly now \o/

Ocramius commented on src/MiddlewareListener.php:
```diff
@@ -64,7 +80,15 @@ public function onDispatch(MvcEvent $event)
+                $psr7Request,
+                $psr7ResponsePrototype,
+                function (PsrServerRequestInterface $request, PsrResponseInterface $response) {
+                    throw new ReachedFinalHandlerException(
```
@weierophinney are you OK with moving this to a static named constructor instead? Asking because consistency...

weierophinney replied:
I am, but I'm not convinced it's needed, as there's a static message with no dynamic substitutions, and no additional behavior.

weierophinney replied:
One potential refactor I see here is to instead define `ReachedFinalHandlerException` as follows:

```php
class ReachedFinalHandlerException extends RuntimeException
{
    public static function createFinalHandler()
    {
        return new self('Reached the final handler for middleware pipe - check the pipe configuration');
    }

    public function __invoke()
    {
        throw $this;
    }
}
```

This would then allow you to do the following above:

```php
$return = $pipe($psr7Request, $psr7Response, ReachedFinalHandlerException::createFinalHandler());
```

Not sure if that's abusing inheritance and SRP, though. :grin:

asgrim replied:
Yikes, I'm not really sure about the exception itself being the final handler, although I gotta say it's quite clever... does feel a bit wrong. I'll just refactor to named constructor for now and if you want to change it to be the actual final handler, I'm happy to :)

Ocramius commented on src/MiddlewareListener.php:
```diff
@@ -47,15 +50,28 @@ public function onDispatch(MvcEvent $event)
+
+        $middlewareName = 'noMiddlewarePiped';
+        $middlewareToBePiped = null;
+        foreach ($middlewaresToBePiped as $middlewareToBePiped) {
```
Segregate to a private method.

Ocramius replied:
Probably should just return `$pipe` in the private method.

asgrim replied:
Done

Ocramius commented on src/MiddlewareListener.php:
```diff
@@ -47,15 +50,28 @@ public function onDispatch(MvcEvent $event)
+        $pipe = new MiddlewarePipe();
+        $pipe->setResponsePrototype($psr7ResponsePrototype);
+
+        $middlewaresToBePiped = !is_array($middleware) ? [$middleware] : $middleware;
```
Don't negate: invert the ternary please

asgrim replied:
Done

Ocramius commented on src/MiddlewareListener.php:
```diff
@@ -47,15 +50,28 @@ public function onDispatch(MvcEvent $event)
+        $middlewareName = 'noMiddlewarePiped';
+        $middlewareToBePiped = null;
+        foreach ($middlewaresToBePiped as $middlewareToBePiped) {
+            $middlewareName = is_string($middlewareToBePiped) ? $middlewareToBePiped : get_class($middlewareToBePiped);
```
Because [`get_class()` is stupid](https://3v4l.org/5O4QA) this check is insufficient

asgrim replied:
Didn't change this from the original code...

Ocramius commented on src/MiddlewareListener.php:
```diff
@@ -47,15 +50,28 @@ public function onDispatch(MvcEvent $event)
+        foreach ($middlewaresToBePiped as $middlewareToBePiped) {
+            $middlewareName = is_string($middlewareToBePiped) ? $middlewareToBePiped : get_class($middlewareToBePiped);
+
+            if (is_string($middlewareToBePiped) && $serviceManager->has($middlewareToBePiped)) {
```
`is_string()` check is duplicate

asgrim replied:
Not sure what to do about it; it's just the flexibility around using a string as the middleware (load from container), or a callable (not ideal, but allowed)

Ocramius commented on src/MiddlewareListener.php:
```diff
@@ -47,15 +50,28 @@ public function onDispatch(MvcEvent $event)
+            if (is_string($middlewareToBePiped) && $serviceManager->has($middlewareToBePiped)) {
+                $middlewareToBePiped = $serviceManager->get($middlewareToBePiped);
+            }
+            if (! is_callable($middlewareToBePiped)) {
```
Should probably also **NOT** be a string?

Ocramius replied:
An example is an incompatible function name being passed in.

asgrim replied:
If it's a string, it won't be a string any more. Also just following same logic that was already there...

Ocramius commented on src/MiddlewareListener.php:
```diff
@@ -75,7 +99,9 @@ public function onDispatch(MvcEvent $event)
             $event->setController($middlewareName);
-            $event->setControllerClass(get_class($middleware));
+            if (null !== $middlewareToBePiped) {
+                $event->setControllerClass(get_class($middlewareToBePiped));
```
You always assume that the last piped middleware is the "responding" one here... Probably needs documenting/commenting

asgrim replied:
Removed these as there's no longer just one middleware, so doesn't make sense. I guess this causes a definite BC break which is kinda rubbish though?

Ocramius commented on src/MiddlewareListener.php:
```diff
@@ -47,15 +50,28 @@ public function onDispatch(MvcEvent $event)
+        $middlewaresToBePiped = !is_array($middleware) ? [$middleware] : $middleware;
+
+        $middlewareName = 'noMiddlewarePiped';
+        $middlewareToBePiped = null;
```
Can an empty pipeline be considered valid?

asgrim replied:
I don't see why not; you'll just reach the final handler anyway...

asgrim committed:
Extract logic to create middleware pipe into private method

asgrim committed:
Do not allow a null middleware

asgrim commented:
@Ocramius some changes applied as requested.

asgrim committed:
Remove null coalesce operator because still need to support PHP 5.6

asgrim commented:
*shakes fist at PHP 5.6 support*

weierophinney committed:
Merge branch 'hotfix/232'

Close #232

weierophinney commented on composer.json:
```diff
@@ -27,7 +27,8 @@
         "fabpot/php-cs-fixer": "1.7.*",
-        "phpunit/phpunit": "^4.5"
+        "phpunit/phpunit": "^4.5",
+        "zendframework/zend-stratigility": "^2.0"
```
This should likely be `^2.0.1`, but I can take care of that during merge.

weierophinney commented on src/MiddlewareListener.php:
```diff
@@ -74,8 +97,6 @@ public function onDispatch(MvcEvent $event)
             $event->setName(MvcEvent::EVENT_DISPATCH_ERROR);
             $event->setError($application::ERROR_EXCEPTION);
-            $event->setController($middlewareName);
-            $event->setControllerClass(get_class($middleware));
```
Why are these lines removed?

asgrim replied:
There is potentially no longer just "one" middleware name... how do we decide which is the failed one? First? Last? IIRC I don't think there's any way (outside of the exception bubble here)

weierophinney replied:
Ah, right - that makes sense, then.

weierophinney commented on src/MiddlewareListener.php:
```diff
@@ -96,6 +117,42 @@ public function onDispatch(MvcEvent $event)
+            if (is_string($middlewareToBePiped) && $serviceLocator->has($middlewareToBePiped)) {
+                $middlewareToBePiped = $serviceLocator->get($middlewareToBePiped);
+            }
+            if (! is_callable($middlewareToBePiped)) {
```
This needs to change to be:

```php
if (! $middlewareToBePiped instanceof \Http\Interop\ServerMiddleware\MiddlewareInterface
    && ! is_callable($middlewareToBePiped)
)
```

as zend-stratigility 2.0 allows either of those types. Testing only for `is_callable()` means that http-interop middleware cannot be used!

weierophinney replied:
Once you make those changes, you'll also need to add tests for piping http-interop middleware.

weierophinney commented:
Also, @asgrim — You'll need to rebase against latest develop branch, as I pushed a bunch of QA tooling changes earlier (PHPUnit 5.7/6.0 support, in particular, as well as zend-coding-standards integration).

asgrim committed:
Added support for HttpInterop style middlewares

asgrim committed:
Refactored ReachedFinalHandlerException to use named constructor

asgrim committed:
Fixed typo in phpdoc for MvcEvent

asgrim committed:
Switch MiddlewareNotCallableExceptionTest assertions to be invoked on instance for consistency

asgrim committed:
Fixed incorrectly rebased composer dependencies

asgrim committed:
Use namespaced PHPUnit import

asgrim committed:
Use expectException instead of deprecated setExpectedException

asgrim commented on test/MiddlewareListenerTest.php:
```diff
@@ -88,6 +92,38 @@ public function testSuccessfullyDispatchesMiddleware()
+    {
+        $expectedOutput = uniqid('expectedOutput', true);
+
+        $event = $this->createMvcEvent('path', new class($expectedOutput) implements MiddlewareInterface {
```
bah, doesn't work on PHP 5.6 - I thought `develop` branches were min PHP 7.1 now? Maybe I didn't read properly?

samsonasik committed:
Fixes @return DefaultRenderingStrategy typo at HttpDefaultRenderingStrategyFactory

asgrim committed:
Use mocks instead of anon classes for middlewares

asgrim commented:
@weierophinney all updated, thanks for review!

weierophinney committed:
Merge pull request #237 from samsonasik/@return-rendering-strategy

Fixes @param & @return DefaultRenderingStrategy typo at HttpDefaultRenderingStrategyFactory

weierophinney committed:
Added CHANGELOG for #237

weierophinney committed:
Merge branch 'hotfix/237' into develop

Forward port #237

Conflicts:
	src/Service/HttpDefaultRenderingStrategyFactory.php

weierophinney commented on src/MiddlewareListener.php:
```diff
@@ -69,7 +80,13 @@ public function onDispatch(MvcEvent $event)
                 $psr7Request = $psr7Request->withAttribute($key, $value);
             }
-            $return = $middleware($psr7Request, Psr7Response::fromZend($response));
+            $return = $pipe(
```
@asgrim So sorry I didn't notice this earlier, but... If we're pinning to Stratigility 2, we should likely use the http-interop `process()` method here instead of `__invoke()`, which means using a `DelegateInterface` instead of a callable.

May I suggest the following:

```php
$pipe->setResponsePrototype($psr7ResponsePrototype);
$return = $pipe->process($psr7Request, new \Zend\Stratigility\Delegate\CallableDelegateDecorator(
    function (PsrServerRequestInterface $request, PsrResponseInterface $response) {
        throw ReachedFinalHandlerException::create();
    }
));
```

That would accomplish the same as what you have here, and be forwards compatible.

asgrim replied:
Yes, this seems reasonable, though I note that `__invoke` does the wrapping in the `CallableDelegateDecorator` for us, but I have no strong feelings either way. If `__invoke` is eventually going to go, then yes the switch makes sense!

`setResponsePrototype` is already called in `createPipeFromSpec` btw, so I don't think needs calling again (unless I missed something)

weierophinney replied:
Ah, okay, nice.

As noted, using `process()` will be forwards compatible, meaning that when Stratigility adopts a finalized PSR-15, we likely can just add an `||` to the constraint and not need any actual code changes. I figure, do it once. :smile:

weierophinney commented on src/Exception/MiddlewareNotCallableException.php:
```diff
@@ -0,0 +1,38 @@
+
+namespace Zend\Mvc\Exception;
+
+final class MiddlewareNotCallableException extends RuntimeException
```
Since we're allowing either callable or http-interop middleware, perhaps we should change this to `InvalidMiddlewareException` instead?

weierophinney commented on src/MiddlewareListener.php:
```diff
@@ -101,6 +116,42 @@ public function onDispatch(MvcEvent $event)
+        $pipe->setResponsePrototype($responsePrototype);
+        foreach ($middlewaresToBePiped as $middlewareToBePiped) {
+            if (null === $middlewareToBePiped) {
+                throw new \InvalidArgumentException('Middleware name cannot be null');
```
If we rename the exception to `InvalidMiddlewareException`, we could also add another named constructor: `::becauseNull()` (obviously, choose something better!).

weierophinney reviewed (changes requested):
Few last nit-picks, but looks great!

weierophinney reviewed (approved):
Per discussion in Slack, I'll perform the last requested changes now and merge.

Thanks, @asgrim!

weierophinney committed:
Merge pull request #217 from asgrim/modify-middleware-listener-to-use-stratigility-pipe

Modify middleware listener to use stratigility pipe

weierophinney committed:
Rename MiddlewareNotCallableException to InvalidMiddlewareException

Also, adds a `fromNull()` method, allowing it to be used in cases where
a `null` value is provided for middleware. This means that `null`
middleware is handled exactly the same way as any other invalid
middleware type, which required changes to several test expectations.

weierophinney committed:
Use http-interop process()

Instead of `__invoke()`, for forwards compatibility once PSR-15 is
ratified.

weierophinney commented:
Thanks, @asgrim
